mmv1/third_party/terraform/acctest/resource_inventory_reader.go

```diff
@@ -0,0 +1,249 @@
+	"sync"
+
+	"gopkg.in/yaml.v2"
+)
+
+// ResourceYAMLMetadata represents the structure of the metadata files
+type ResourceYAMLMetadata struct {
+	Resource       string `yaml:"resource"`
+	ApiServiceName string `yaml:"api_service_name"`
+	SourceFile     string `yaml:"source_file"`
+}
+
+// Cache structures to avoid repeated file system operations
+var (
+	// Cache for API service names (resourceName -> apiServiceName)
+	apiServiceNameCache = make(map[string]string)
+	// Cache for service packages (resourceType -> servicePackage)
+	servicePackageCache = make(map[string]string)
+	// Flag to track if cache has been populated
+	cachePopulated = false
+	// Mutex to protect cache access
+	cacheMutex sync.RWMutex
+)
+
+// PopulateMetadataCache walks through all metadata files once and populates
+// both the API service name and service package caches for improved performance
+func PopulateMetadataCache() error {
+	cacheMutex.Lock()
+	defer cacheMutex.Unlock()
+
+	// If cache is already populated, we can skip
+	if cachePopulated {
+		return nil
+	}
+
+	baseDir, err := getServicesDir("")
+	if err != nil {
+		return fmt.Errorf("failed to find services directory: %v", err)
+	}
+
+	// Count for statistics
+	apiNameCount := 0
+	servicePkgCount := 0
+
+	// Walk through all service directories once
+	err = filepath.Walk(baseDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return nil // Skip files with errors but continue walking
+		}
+
+		// Look for metadata files
+		if !info.IsDir() && strings.HasPrefix(info.Name(), "resource_") && strings.HasSuffix(info.Name(), "_meta.yaml") {
+			// Read the file
+			content, err := os.ReadFile(path)
+			if err != nil {
+				return nil // Continue to next file
+			}
+
+			// Parse YAML
+			var metadata ResourceYAMLMetadata
+			if err := yaml.Unmarshal(content, &metadata); err != nil {
+				return nil // Continue to next file
+			}
+
+			// Skip if resource is empty
+			if metadata.Resource == "" {
+				return nil
+			}
+
+			// Store API service name in cache
+			if metadata.ApiServiceName != "" {
+				apiServiceNameCache[metadata.Resource] = metadata.ApiServiceName
+				apiNameCount++
+			}
+
+			// Extract and store service package in cache
+			pathParts := strings.Split(path, string(os.PathSeparator))
+			servicesIndex := -1
+			for i, part := range pathParts {
+				if part == "services" {
+					servicesIndex = i
+					break
+				}
+			}
+
+			if servicesIndex >= 0 && len(pathParts) > servicesIndex+1 {
+				servicePackage := pathParts[servicesIndex+1] // The part after "services"
+				servicePackageCache[metadata.Resource] = servicePackage
+				servicePkgCount++
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("error walking directory: %v", err)
+	}
+
+	// Mark cache as populated
+	cachePopulated = true
+
+	return nil
+}
+
+// GetCacheStats returns statistics about the metadata cache
+func GetCacheStats() (apiServiceNameCount int, servicePackageCount int) {
```

Comment:
I don't see this function is called anywhere.

Reply:
removed